Game/Source/GDKTestGyms/Private/LatencyTracer.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "LatencyTracer.h"
+
+#include "Engine/World.h"
+#include "SpatialGDKSettings.h"
+
+DEFINE_LOG_CATEGORY(LogLatencyTracer);
+
+namespace LatencyTracerInternal
+{
+	SpatialGDK::SpatialEventTracer* GetEventTracerFromNetDriver(const UWorld* World)
+	{
+		if (USpatialNetDriver* NetDriver = Cast<USpatialNetDriver>(World->GetNetDriver()))
+		{
+			return NetDriver->Connection->GetEventTracer();
+		}
+
+		return nullptr;
+	}
+}
+
+ULatencyTracer* ULatencyTracer::GetOrCreateLatencyTracer(UObject* WorldContextObject)
+{
+	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::ReturnNull);
+	if (World == nullptr)
+	{
+		World = GWorld;
+	}
+
+	if (UGDKTestGymsGameInstance* GameInstance = World->GetGameInstance<UGDKTestGymsGameInstance>())
+	{
+		if (GameInstance->GetTracer() != nullptr)
+		{
+			return GameInstance->GetTracer();
+		}
+	}
+	UE_LOG(LogLatencyTracer, Error, TEXT("Unable to get the game instance tracer object, creating a new one but results may not be as expected."));
+	return NewObject<ULatencyTracer>();
+}
+
+void ULatencyTracer::InitTracer()
+{
+	// TODO: Also remove references to TraceMetadata + LatencyGameMode metadata id (run id?)
+	const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>();
+	if (Settings->GetEventTracingEnabled() && USpatialStatics::IsSpatialNetworkingEnabled()) // Use the tracing hooked up to GDK internals
+	{
+		InternalTracer = LatencyTracerInternal::GetEventTracerFromNetDriver(GetWorld());
+		if (InternalTracer == nullptr)
+		{
+			UE_LOG(LogLatencyTracer, Warning, TEXT("Full tracing enabled but tracer is not available."));
+		}
+	}
+	else
```

Comment:
So we also construct a LocalTracer when `Settings->GetEventTracingEnabled() == false`?

Reply:
Couldn't we just not create a tracer if `Settings->GetEventTracingEnabled() == false` ?

Reply:
This is the event tracer that is created for native runs I believe. In native we wont set up a tracer on the net driver. @improbable-danny What do you think of my previous comment?

Reply:
The naming is pretty ambiguous in that case, deserves a comment at least.